### ISSUE TITLE
Generating coverage produces errors when testing code that uses ES 6 default/override pattern

### ISSUE THREAD

**Issue by patrickarlt** [bcoe/nyc#264]
I'm testing code that uses the [ES 6 default override pattern](https://gist.github.com/ericelliott/f3c2a53a1d4100539f71) in some code I am testing. When testing regularly with node-tap (which bundles nyc) or ava (which recommends nyc) the tests pass. But when I use nyc to generate coverage reports there are errors that are not present when not generating coverage.

I've created a [minimal reproduction of this issue in this gist](https://gist.github.com/patrickarlt/db48968e267c6d50b08b9459c33d6cb2/). which shows Node 6.2.0 being able to run the tests without coverage (`tap node-tap-test.js` and `ava ava-test.js`) and passing but when running with coverage (`tap test-node-tap.js --coverage-report=text` and `nyc ava ava-test.js`) the tests fail and the parameters padded to the `say(options)` function end up being undefined.

**Comment by JaKXz** [comment]
@patrickarlt I got the ava tests passing with some configuration options and fixing some typos in the `package.json`: https://gist.github.com/JaKXz/f64d82899e00495aba83ddd16d702ee9/revisions

I recommend taking a look at [ava's instructions for `.babelrc` configuration](https://github.com/avajs/ava/blob/master/docs/recipes/babelrc.md) and taking note that you should use `nyc --require babel-register npm run <your test task>` to get the same transpilation for coverage reports.

Let me know if that helps or not :)

**Comment by novemberborn** [comment]
There are various issues with ES6 code and Istanbul, the library nyc uses to collect coverage data and prepare the reports: #215, #198, #239.

The tl;dr is that Istanbul doesn't support the new syntax, so you need to transpile to ES5, which is @JaKXz's suggestion.

Unfortunately that doesn't always work, since nyc now needs source maps in order to get the correct data. Source maps aren't always accurate, and then the resulting coverage data may be internally inconsistent, leading to crashes when generating reports.

There's a few things we can do to reduce the incidence of crashes (namely to throw away inconsistent data), but then there'll be less coverage.

The better approach would be to switch to an instrumenter that understands ES6. I've been planning to work on that but haven't had the time.

In the meantime try transpiling to ES5 and hope for the best.

**Comment by bcoe** [comment]
@patrickarlt give this a spin:

https://github.com/bcoe/nyc#use-babel-plugin__coverage__-for-better-es6es7-support

we've released better support for ES6/ES6 coverage using the `__coverage__` instrumentation library.
